Adopt more smart pointers in ImageLoader

rwlbuis opened the pull request:
#### b39fe58ededffb0ebbcff8308b27849ae1466200
<pre>
Adopt more smart pointers in ImageLoader
<a href="https://bugs.webkit.org/show_bug.cgi?id=276605">https://bugs.webkit.org/show_bug.cgi?id=276605</a>

Reviewed by Ryosuke Niwa.

Adopt smart pointers for ImageLoader based on [alpha.webkit.UncountedCallArgsChecker] warnings.

* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::activeServiceWorker const): Deleted.
* Source/WebCore/dom/ScriptExecutionContext.h:
(WebCore::ScriptExecutionContext::activeServiceWorker const):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::canReuseFromListOfAvailableImages):
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::notifyFinished):
(WebCore::ImageLoader::decode):
(WebCore::ImageLoader::resetLazyImageLoading):
(WebCore::ImageLoader::shouldIgnoreCandidateWhenLoadingFromArchive const):
* Source/WebCore/loader/ImageLoader.h:
(WebCore::ImageLoader::protectedDocument):

Canonical link: <a href="https://commits.webkit.org/282159@main">https://commits.webkit.org/282159@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/77b6048b34adc9119abcff24918c6790859e83b4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62127 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41481 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14719 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66107 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12672 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64246 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49167 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13012 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50056 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8764 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65196 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38518 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53808 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30871 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35180 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11060 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11603 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56940 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11365 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67836 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6070 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/11131 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/57433 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6096 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53771 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57668 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13832 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4997 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37281 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38365 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39461 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38110 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/EventSender.h:
```diff
@@ -121,7 +121,7 @@ template<typename T, typename WeakPtrImpl> void EventSender<T, WeakPtrImpl>::dis
         if (!weakSender)
             continue;
         auto sender = std::exchange(weakSender, nullptr);
-        if (!page || sender->document().page() == page)
```
page() is definitely a trivial getter so we shouldn't have to use protectedDocument here.

rwlbuis replied:
Yes, not sure what I was thinking :) Fixed.

rniwa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -245,7 +245,7 @@ void ImageLoader::updateFromElement(RelevantMutation relevantMutation)
             bool autoLoadOtherImages = cachedResourceLoader->autoLoadImages();
             cachedResourceLoader->setAutoLoadImages(false);
-            RefPtr page = m_element->document().page();
+            RefPtr page = protectedDocument()->page();
```
Again, page() is a trivial getter so we shouldn't need to use `protectedDocument`.

rwlbuis replied:
Fixed.

rniwa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -359,7 +359,7 @@ CachedResourceHandle<CachedImage> ImageLoader::protectedImage() const
 void ImageLoader::updateFromElementIgnoringPreviousErrorToSameValue()
 {
-    if (!m_image || !m_image->allowsCaching() || !m_failedLoadURL.isEmpty() || element().document().activeServiceWorker()) {
+    if (!m_image || !m_image->allowsCaching() || !m_failedLoadURL.isEmpty() || protectedDocument()->activeServiceWorker()) {
```
This is silly. We should just move the definition of `ScriptExecutionContext::activeServiceWorker` to the header file.

rwlbuis replied:
Done.

rniwa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -598,7 +598,7 @@ void ImageLoader::dispatchPendingBeforeLoadEvent()
         return;
     if (!m_image)
         return;
-    if (!element().document().hasLivingRenderTree())
```
hasLivingRenderTree is a trivial getter so we shouldn't need to do use protectedDocument here.

rwlbuis replied:
Fixed.

rniwa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -613,7 +613,7 @@ void ImageLoader::dispatchPendingLoadEvent()
     if (!m_image)
         return;
     m_hasPendingLoadEvent = false;
-    if (element().document().hasLivingRenderTree())
```
Ditto.

rwlbuis replied:
Fixed.

rniwa commented on Source/WebCore/loader/ImageLoader.cpp:
```diff
@@ -627,7 +627,7 @@ void ImageLoader::dispatchPendingErrorEvent()
         return;
     m_hasPendingErrorEvent = false;
     loadEventSender().cancelEvent(*this, eventNames().errorEvent);
-    if (element().document().hasLivingRenderTree())
```
Ditto.

rwlbuis replied:
Fixed.

rniwa reviewed (changes requested):
Some of the smart pointer uses in this PR is excessive.

rwlbuis committed:
Adopt more smart pointers in ImageLoader
https://bugs.webkit.org/show_bug.cgi?id=276605

Reviewed by Ryosuke Niwa.

Adopt smart pointers for ImageLoader based on [alpha.webkit.UncountedCallArgsChecker] warnings.

* Source/WebCore/dom/ScriptExecutionContext.cpp:
(WebCore::ScriptExecutionContext::activeServiceWorker const): Deleted.
* Source/WebCore/dom/ScriptExecutionContext.h:
(WebCore::ScriptExecutionContext::activeServiceWorker const):
* Source/WebCore/loader/ImageLoader.cpp:
(WebCore::canReuseFromListOfAvailableImages):
(WebCore::ImageLoader::updateFromElement):
(WebCore::ImageLoader::notifyFinished):
(WebCore::ImageLoader::decode):
(WebCore::ImageLoader::resetLazyImageLoading):
(WebCore::ImageLoader::shouldIgnoreCandidateWhenLoadingFromArchive const):
* Source/WebCore/loader/ImageLoader.h:
(WebCore::ImageLoader::protectedDocument):

Canonical link: https://commits.webkit.org/282159@main
